optaplanner-examples/src/main/java/org/optaplanner/examples/taskassigning/solver/score/TaskAssigningConstraintProvider.java

```diff
@@ -0,0 +1,80 @@
+                .filter(task -> task.getMissingSkillCount() > 0)
+                .penalize("No missing skills",
+                        BendableScore.ofHard(BENDABLE_SCORE_HARD_LEVELS_SIZE, BENDABLE_SCORE_SOFT_LEVELS_SIZE, 0, 1),
+                        Task::getMissingSkillCount);
+    }
+
+    private static Constraint criticalPriorityBasedTaskEndTime(ConstraintFactory constraintFactory) {
+        return getInitializedTaskUniStream(constraintFactory)
+                .filter(task -> task.getPriority() == Priority.CRITICAL)
+                .penalize("Critical priority task end time",
+                        BendableScore.ofSoft(BENDABLE_SCORE_HARD_LEVELS_SIZE, BENDABLE_SCORE_SOFT_LEVELS_SIZE, 0, 1),
+                        Task::getEndTime);
+    }
+
+    private static Constraint minimizeMakespan(ConstraintFactory constraintFactory) {
+        return getInitializedTaskUniStream(constraintFactory)
+                .filter(task -> task.getNextTask() == null)
+                .penalize("Minimize makespan, latest ending employee first",
+                        BendableScore.ofSoft(BENDABLE_SCORE_HARD_LEVELS_SIZE, BENDABLE_SCORE_SOFT_LEVELS_SIZE, 1, 1),
+                        task -> task.getEndTime() * task.getEndTime());
+        /*
+         * TODO ^^^ Some constraints need exponential weighting. perhaps we should have dynamic
+         *  constraint weights, inferred from the entity?
+         */
+    }
+
+    private static Constraint majorPriorityTaskEndTime(ConstraintFactory constraintFactory) {
+        return getInitializedTaskUniStream(constraintFactory)
+                .filter(task -> task.getPriority() == Priority.MAJOR)
+                .penalize("Major priority task end time",
+                        BendableScore.ofSoft(BENDABLE_SCORE_HARD_LEVELS_SIZE, BENDABLE_SCORE_SOFT_LEVELS_SIZE, 2, 1),
+                        Task::getEndTime);
+    }
+
+    private static Constraint minorPriorityTaskEndTime(ConstraintFactory constraintFactory) {
+        return getInitializedTaskUniStream(constraintFactory)
+                .filter(task -> task.getPriority() == Priority.MINOR)
+                .penalize("Minor priority task end time",
+                        BendableScore.ofSoft(BENDABLE_SCORE_HARD_LEVELS_SIZE, BENDABLE_SCORE_SOFT_LEVELS_SIZE, 3, 1),
+                        Task::getEndTime);
+    }
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                noMissingSkills(constraintFactory),
+                minimizeMakespan(constraintFactory),
+                /*
+                 * TODO if we had a way of determining constraint weight from the entity, different weight for different
+                 *  priority, then the following 3 constraints could have been one constraint, with associated
+                 *  performance benefits.
+                 */
+                criticalPriorityBasedTaskEndTime(constraintFactory),
```

Comment:
@ge0ffrey What do you think about this comment? This really seems like a useful addition to the API.

Instead of three constraints:

- filter(task.getPriority() == Priority.CRITICAL)) => add penalty X
- filter(task.getPriority() == Priotiry.MAJOR)) => add penalty Y
- filter(task.getPriority() == Priority.MINOR)) = > add penalty Z

We could have just one constraint:

- task -> if (Priority.CRITICAL) => add penalty X, else if (Priority.MAJOR) => ...

Reply:
Either having a "free for all" penalize would do the job, or - in this particular case - something like `penalizeBendable` would help too. Currently, the result of `ToIntFunction<Task>` is just multiplied by each bendable score weight level passed in the previous parameter.

Reply:
It's a good idea - but I 'd handle this as a separate issue. This change would need to be applied to the scoreRules.drl too. Right now, I 'd like have an apples to apples comparison for drl vs constraintProvider